File: day06/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	count := 0
	s := map[byte]struct{}{}

	count2 := 0
	k := map[byte]int{}
	people := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			count += len(s)
			s = map[byte]struct{}{}
			for _, v := range k {
				if v == people {
					count2++
				}
			}
			k = map[byte]int{}
			people = 0
			continue
		}
		people++
		for _, i := range []byte(line) {
			s[i] = struct{}{}
			if _, ok := k[i]; !ok {
				k[i] = 0
			}
			k[i]++
		}
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	count += len(s)

	for _, v := range k {
		if v == people {
			count2++
		}
	}

	fmt.Println("Part 1:", count)
	fmt.Println("Part 2:", count2)
}

File: day06/src/main.rs
use std::collections::HashSet;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut count = 0;
    let mut set = HashSet::new();

    let mut count2 = 0;
    let mut set2 = HashSet::new();
    let mut is_init = false;

    for line in reader.lines() {
        let line = line?;
        if line == "" {
            count += set.len();
            set = HashSet::new();

            count2 += set2.len();
            is_init = false;
            continue;
        }
        let chars = line.chars().collect::<HashSet<char>>();
        set = set.union(&chars).cloned().collect();

        if is_init {
            set2 = set2.intersection(&chars).cloned().collect();
        } else {
            is_init = true;
            set2 = chars;
        }
    }

    count += set.len();
    count2 += set2.len();

    println!("Part 1: {}", count);
    println!("Part 2: {}", count2);

    Ok(())
}
